File: stacks/rails-app/app/controllers/products_controller.rb
class ProductsController < ApplicationController
  def create
    product = Product.create!(params.permit(:name, :price))
    render json: serialize_product(product), status: :created
  end

  def index
    render json: Product.order(:id).map { |product| serialize_product(product) }
  end

  def show
    render json: serialize_product(Product.find(params[:id]))
  end

  private

  def serialize_product(product)
    {
      id: product.id,
      name: product.name,
      price: product.price
    }
  end
end

File: stacks/laravel-app/app/Http/Controllers/ProductsController.php
<?php

namespace App\Http\Controllers;

use App\Models\Product;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Validator;

class ProductsController extends Controller
{
    public function store(Request $request): JsonResponse
    {
        $validator = Validator::make($request->all(), [
            'name' => ['required', 'string'],
            'price' => ['required', 'numeric', 'gt:0'],
        ]);

        if ($validator->fails()) {
            return response()->json(['detail' => $validator->errors()->first()], 422);
        }

        $validated = $validator->validated();
        $product = Product::create($validated);

        return response()->json($product, 201);
    }

    public function index(): JsonResponse
    {
        return response()->json(Product::query()->orderBy('id')->get());
    }

    public function show(int $id): JsonResponse
    {
        $product = Product::query()->find($id);

        if (! $product) {
            return response()->json(['detail' => 'Product not found'], 404);
        }

        return response()->json($product);
    }
}
